src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java

```diff
@@ -117,7 +108,18 @@ public final void checkout(Run<?,?> run, FilePath workspace, TaskListener listen
                 }
                 }
             }
-            scm.checkout(run, launcher, workspace, listener, changelogFile, baseline);
+            File changelogFile = null;
+            synchronized (run) {
+                if (changelog) {
+                    for (int i = 0; ; i++) {
+                        changelogFile = new File(run.getRootDir(), "changelog" + i + ".xml");
+                        if (!changelogFile.exists()) {
+                            break;
+                        }
+                    }
+                }
+                scm.checkout(run, launcher, workspace, listener, changelogFile, baseline);
```

Comment:
Seems undesirable to have this in a synchronized block. Maybe a better approach would be to only synchronize while looking for an available filename, touch that new file in the synchronized block to claim it so concurrent callers do not use it, and then the checkout itself would not need to be synchronized. I don't think there are enough tests in this plugin to know whether or not that would work, so would be good to test out any patches with downstream plugins (maybe workflow-multibranch?).

Reply:
> touch that new file in the synchronized block to claim it

I thought of that. The risk is that something goes wrong during checkout and you leave behind an empty (and thus malformed) XML file in the build directory, leading to similar confusing errors. Perhaps possible to defend against this by deleting the file if `checkout` fails, but I think there could still be race conditions if Jenkins is restarted etc.

Perhaps we could introduce a changelog count field into `MultiSCMRevisionState` or something like that?

Reply:
Or your idea in JIRA may pan out: pass a temporary file for `changelogFile` and only rename to `changelogNNN.xml` if `checkout` succeeds, picking a unique name at that time.

Reply:
> pass a temporary file for changelogFile and only rename to changelogNNN.xml if checkout succeeds, picking a unique name at that time

I went with that approach in 7286a18, but it still needs to be tested.